### PR TITLE
feature/KAA-306

### PR DESCRIPTION
**Pull request by akokhanovskyi**
Basic restructuring. New structure to be used in implementation going forward.
Reviewers: @andreypanasenko and @dyosick

### CODE REVIEW

File: client/client-multi/client-c/src/kaa_platform_utils.c [hunk 1]
```diff
@@ -67,6 +67,22 @@ kaa_error_t kaa_platform_message_write(kaa_platform_message_writer_t* writer
 
 
 
+kaa_error_t kaa_platform_message_write_alignment(kaa_platform_message_writer_t* writer)
+{
+    KAA_RETURN_IF_NIL(writer, KAA_ERR_BADPARAM);
+
+    size_t alignment_size = KAA_ALIGNMENT - ((writer->current - writer->begin) % KAA_ALIGNMENT);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa_platform_utils.c]: The size of the alignment is calculated by this formula:
(KAA_ALIGNMENT - (size % KAA_ALIGNMENT)) % KAA_ALIGNMENT

Assume, size = 16 and KAA_ALIGNMENT = 4. In this case no padding should be added.
Current formula: 4 - (16 % 4) = 4
Proposed formula: (4 - (16 % 4)) % 4 = 0

File: client/client-multi/client-c/src/kaa_platform_utils.c [hunk 2]
```diff
@@ -67,6 +67,22 @@ kaa_error_t kaa_platform_message_write(kaa_platform_message_writer_t* writer
...
+
+    if (alignment_size && (writer->current + alignment_size <= writer->end)) {
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa_platform_utils.c]: Is it an error if the alignment_size variable has 0 value?

File: client/client-multi/client-c/src/kaa_logging.c [hunk 3]
```diff
@@ -149,36 +143,38 @@ static void update_storage(kaa_log_collector_t *self)
...
-    KAA_RETURN_IF_NIL(self->log_storage.add_log_record, KAA_ERR_NOT_INITIALIZED);
+    KAA_RETURN_IF_NIL(self->log_storage, KAA_ERR_NOT_INITIALIZED);
 
     KAA_LOG_DEBUG(self->logger, KAA_ERR_NONE, "Adding new log record {%p}", entry);
 
-    kaa_log_entry_t record = { NULL, 0 };
-
-    record.record_size = entry->get_size(entry);
+    kaa_log_record_t record = { NULL, entry->get_size(entry) };
+    KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Record size is %d", record.size);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa_logging.c]: The type of the size field (struct kaa_log_record_t) has the size_t type, so %zu specifier should be used.

File: client/client-multi/client-c/src/kaa_logging.c [hunk 4]
```diff
@@ -34,7 +34,6 @@
 
 #define KAA_LOGGING_RECEIVE_UPDATES_FLAG   0x01
 #define KAA_MAX_PADDING_LENGTH             3
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa_logging.c]: Maybe better to use the (KAA_ALIGNMENT - 1) value for the KAA_MAX_PADDING_LENGTH?

File: client/client-multi/client-c/src/kaa_logging.c [hunk 5]
```diff
@@ -220,37 +220,51 @@ kaa_error_t kaa_logging_request_serialize(kaa_log_collector_t *self, kaa_platfor
...
-    char *records_count_p = writer->current; // pointer to the records count. Will be filled later.
-    writer->current += sizeof(uint16_t);
+    *((uint16_t *) tmp_writer.current) = KAA_HTONS(self->log_bucket_id);
+    tmp_writer.current += sizeof(uint16_t);
+    char *records_count_p = tmp_writer.current; // Pointer to the records count. Will be filled in later.
+    tmp_writer.current += sizeof(uint16_t);
 
-    ssize_t remaining_size = self->log_properties.max_log_bucket_size;
-    KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Extracting log records... (Bucket size is %zu)", remaining_size);
+    ssize_t remaining_size = self->max_log_bucket_size < (tmp_writer.end - tmp_writer.current)
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa_logging.c]: Why the remaining size isn't set to the max bucket size when the size of the unused space in a buffer matches to it?

File: client/client-multi/client-c/src/kaa_logging.c [hunk 6]
```diff
@@ -220,37 +220,51 @@ kaa_error_t kaa_logging_request_serialize(kaa_log_collector_t *self, kaa_platfor
...
-        writer->current += sizeof(uint32_t);
-
-        if (kaa_platform_message_write_aligned(writer, entry.record_data, entry.record_size)) {
-            KAA_LOG_ERROR(self->logger, KAA_ERR_WRITE_FAILED, "Failed to write the log record buffer {%p}", entry.record_data);
-            if (self->log_storage.upload_failed)
-                self->log_storage.upload_failed(self->log_storage.context, self->log_bucket_id);
-            return KAA_ERR_WRITE_FAILED;
+
+    while (!error) {
+        size_t record_len = 0;
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa_logging.c:234]: The record_len variable may be moved to the upper level.

File: client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+
+
+kaa_error_t ext_log_storage_allocate_log_record_buffer(ext_log_storage_t *self, kaa_log_record_t *record)
+{
+    KAA_RETURN_IF_NIL2(record, record->size, KAA_ERR_BADPARAM);
+
+    record->data = (char *) KAA_MALLOC(record->size * sizeof(char));
+    return KAA_ERR_NONE;
```
Comment (ghost) [inline client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c:94]: KAA_ERR_NOMEM should be returned if malloc has failed.

File: client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+
+
+kaa_error_t ext_log_storage_write_next_record(ext_log_storage_t *self, char *buffer, size_t buffer_len
+        , uint16_t bucket_id, size_t *record_len)
+{
+    KAA_RETURN_IF_NIL5(self, buffer, buffer_len, bucket_id, record_len, KAA_ERR_BADPARAM);
+
+    uint16_t zero_bucket_id = 0;
+    kaa_list_t *record_position = kaa_list_find_next(self->logs, &logs_list_find_by_bucket_id, &zero_bucket_id);
```
Comment (ghost) [inline client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c]: To reduce overhead I propose to remember the position of the unmarked record and use it (if not null) on the next call.

File: client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+
+kaa_error_t ext_log_storage_shrink_by_size(ext_log_storage_t *self, size_t size)
+{
+    KAA_RETURN_IF_NIL(self, KAA_ERR_BADPARAM);
+    size_t removed_size = 0;
+
+    while ((removed_size < size) && self->occupied_size) {
+        // May delete records already marked with bucket_id. C'est la vie...
+        size_t record_size = ((ext_log_record_t *)kaa_list_get_data(self->logs))->size;
```
Comment (ghost) [inline client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c:224]: The record size variable may be moved to the upper level.

**Comment by andreypanasenko** [comment]
Nothing to add after Denis :)
LGTM

File: client/client-multi/client-c/src/kaa_platform_utils.c (hunk 1 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_platform_utils.c]: Agree

File: client/client-multi/client-c/src/kaa_platform_utils.c (hunk 2 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_platform_utils.c]: My bad

File: client/client-multi/client-c/src/kaa_logging.c (hunk 3 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_logging.c]: ok

File: client/client-multi/client-c/src/kaa_logging.c (hunk 4 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_logging.c]: Agree

File: client/client-multi/client-c/src/kaa_logging.c (hunk 5 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_logging.c]: A very tiny optimization? :+1:

File: client/client-multi/client-c/src/kaa_logging.c (hunk 6 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_logging.c:234]: For what reason?

File: client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c (hunk 7 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c:94]: :+1:

File: client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c (hunk 9 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c:224]: Why?

File: client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c (hunk 8 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/ext_log_storage_memory.c]: I thought about this but did not want to overcomplicate the reference implementation. Anyway, this is fairly simple. Added.

**Pull request closed without merging**